handler_test.go

```diff
@@ -0,0 +1,126 @@
+	ping := rerpc.NewHandler(
+		"internal.ping.v1test.PingService.Ping", // fully-qualified protobuf method
+		"internal.ping.v1test.PingService",      // fully-qualified protobuf service
+		"internal.ping.v1test",                  // fully-qualified protobuf package
+		rerpc.Func(func(ctx context.Context, req proto.Message) (proto.Message, error) {
+			return &pingpb.PingResponse{}, nil
+		}),
+		rerpc.ReadMaxBytes(readMaxBytes),
+	)
+
+	t.Run("read_max_bytes", func(t *testing.T) {
+
+		t.Run("twirp/json", func(t *testing.T) {
+			probe := `{"number":"42", "msg": "padding"}`
+			assert.Equal(t, len(probe), readMaxBytes+1, "probe size")
+			r, err := http.NewRequest(
+				http.MethodPost,
+				"rerpc.io/internal.ping.v1test.PingService/Ping",
+				strings.NewReader(probe),
+			)
+			assert.Nil(t, err, "create request")
+			r.Header.Set("Content-Type", rerpc.TypeJSON)
+
+			recorder := httptest.NewRecorder()
+			ping.Serve(recorder, r, &pingpb.PingRequest{})
+			response := recorder.Result()
+
+			assert.Equal(t, response.StatusCode, http.StatusBadRequest, "HTTP status code")
+			expected := &twirp.Status{
+				Code:    "malformed",
+				Message: "can't unmarshal JSON body",
+			}
+			got := &twirp.Status{}
+			contents, err := io.ReadAll(response.Body)
+			assert.Nil(t, err, "read response body")
+			assert.Nil(t, json.Unmarshal(contents, got), "unmarshal JSON")
+			assert.Equal(t, got, expected, "unmarshaled Twirp status")
+		})
+
+		t.Run("twirp/protobuf", func(t *testing.T) {
+			padding := "padding                      "
+			probeBytes, err := proto.Marshal(&pingpb.PingRequest{Number: 42, Msg: padding})
+			assert.Nil(t, err, "marshal request")
+			assert.Equal(t, len(probeBytes), readMaxBytes+1, "probe size")
+			r, err := http.NewRequest(
+				http.MethodPost,
+				"rerpc.io/internal.ping.v1test.PingService/Ping",
+				bytes.NewReader(probeBytes),
+			)
+			assert.Nil(t, err, "create request")
+			r.Header.Set("Content-Type", rerpc.TypeProtoTwirp)
+
+			recorder := httptest.NewRecorder()
+			ping.Serve(recorder, r, &pingpb.PingRequest{})
+			response := recorder.Result()
+
+			assert.Equal(t, response.StatusCode, http.StatusBadRequest, "HTTP status code")
+			expected := &twirp.Status{
+				Code:    "malformed",
+				Message: "can't unmarshal Twirp protobuf body",
+			}
+			got := &twirp.Status{}
+			contents, err := io.ReadAll(response.Body)
+			assert.Nil(t, err, "read response body")
+			assert.Nil(t, json.Unmarshal(contents, got), "unmarshal JSON")
+			assert.Equal(t, got, expected, "unmarshaled Twirp status")
+		})
+
+		t.Run("grpc", func(t *testing.T) {
+			padding := "padding                      "
+			probeBytes, err := proto.Marshal(&pingpb.PingRequest{Number: 42, Msg: padding})
+			assert.Nil(t, err, "marshal request")
+			assert.Equal(t, len(probeBytes), readMaxBytes+1, "probe size")
+			r, err := http.NewRequest(
+				http.MethodPost,
+				"rerpc.io/internal.ping.v1test.PingService/Ping",
+				bytes.NewReader(probeBytes),
```

Comment:
We should use the generated client here - this isn't actually a valid gRPC request. gRPC request bodies are (potentially) a series of messages, and each one has a 5-byte prefix. This request is missing the prefix, and there's no exported function to add one.

The test passes, but it shouldn't.

Reply:
I saw the change in https://github.com/rerpc/rerpc/commit/b8e87af950a4f95cb02e8acdbbc48629c9f0aae6 — glad I caught that one. I completely forgot about gRPC's own framing.